### PR TITLE
Initial draft of Licensing content

### PR DESCRIPTION
**Pull request by swinslow**
This is a work-in-progress draft of the classes and other content for the Licensing profile. This should not be merged at present, in particular because:

* several placeholder files are empty
* the structure and details for the `*Exception` classes will likely be dependent on the outcome of [the pending change proposal](https://github.com/spdx/change-proposal/issues/4)
* still need to understand better how this as an optional "profile" ties into inserting these properties into classes defined in other "profiles" (such as Core and Software)

Submitting here as a draft to get greater visibility for the community into progress on this.

Signed-off-by: Steve Winslow <steve@swinslow.net>

### CODE REVIEW

**Comment by goneall** [comment]
Thanks @swinslow for getting this started!  I only took a quick glance, but the structure and readability all looked good to me.

I was happy to see some of the fields used in the listed license data repo documented (e.g. `isOsiApproved`).  The extra documentation and standardization will definitely help.  I also like the fact you've structured the full license expressions with nicely detailed descriptions.

When I have some more time, I'll do a more thorough review and provide any feedback/comments.

On you're point above:

> still need to understand better how this as an optional "profile" ties into inserting these properties into classes defined in other "profiles" (such as Core and Software)

this may be a good topic for a future tech call if you're available.

**Review by tsteenbe — CHANGES_REQUESTED** [review]

**Comment by tsteenbe** [comment]
@swinslow Noticed detectedLicenses property is missing... mistake? or still under discussion?

File: model/Licensing/Properties/declaredLicense.md
Comment (swinslow) [reply model/Licensing/Properties/declaredLicense.md:26]: 👍

**Comment by swinslow** [comment]
> @swinslow Noticed detectedLicenses property is missing... mistake? or still under discussion?

@tsteenbe Thanks -- not sure what this means?

* if "detected license" means "license information detected in the package / file / snippet itself", then you would use the `declaredLicense` field.
* or, if "detected license" means "the license you detected by analyzing what's in the package / file / snippet itself as well as other information", then you would use the `concludedLicense` field.

I am not sure what's meant by "detected license" other than one of the above, so I don't think this is either a "mistake" or "still under discussion"  :)

File: model/Licensing/Classes/CopyrightText.md
Comment (swinslow) [reply model/Licensing/Classes/CopyrightText.md:3]: I strongly disagree, either that this is a shortcoming of SPDX 2.x or that this proposal is a simple change.

A copyright holder is not associated directly with a particular license. Rather, a copyright holder is associated with particular content for which they hold the copyright. In an SPDX document, this might be a Package as a whole, or one or more Files within it, or one or more Snippets within those Files. The copyright notice text is a property of the Package / File / Snippet, as is the corresponding license or licenses.

If it is necessary for an SPDX document creator to be more granular about which copyright notice text applies to a particular subunit of code, the appropriate way to handle that would be to model the Artifact at a lower level of granularity. For instance, if you need to talk about which license + copyright text is associated with one part of a File versus another, the right way to handle that would be to create Snippets corresponding to the different parts of the Files with differing license + copyright texts.

I think the proposed change is problematic because it introduces complexity which in my view is not necessary, and which imposes burdens on anyone who is implementing, consuming or creating licensing-related SPDX documents. The current modeling of copyrightText as a simple string which can be associated with any type of Artifact provides sufficient flexibility while being simple to use, consume and create.

**Comment by swinslow** [comment]
With 0aeefef I've now added in the initial drafts of the remaining classes / properties that were previously just empty placeholder files. (Other than the Exception-related ones, which I'm holding off on doing anything further with until there is further movement on the [related change proposal](https://github.com/spdx/change-proposal/issues/4).)

You will see also that there are a number of items marked as `**TBD**` where I have pending questions about how exactly things will work, or open matters to be discussed.

I'm going to go ahead and remove the "Draft" label and will mark this as ready for review. If the tech team is on board with it, I would say let's go ahead and merge this if all looks good, so that folks can begin reviewing and can submit Issues / PRs for further discussion and iteration.

@goneall Thanks -- yes, I'd be happy to try to join part of an upcoming tech team meeting, if there will be a discussion around how the different profiles will relate to the Core / Software elements. And in the meantime I'd definitely welcome your thoughts on the draft profile.

cc @jlovejoy for visibility

**Comment by swinslow** [comment]
@tsteenbe Could you please remove your -1 review?

If your request is to change the way that copyright text is recorded (from a simple string to a new copyright-to-license map), I'm still strongly opposed to that, but in any case I'd ask if it could be discussed in a separate issue (or more likely a [Change Proposal](https://github.com/spdx/change-proposal)) so that this initial draft can be merged for broader visibility and review.

**Comment by zvr** [comment]
@swinslow very briefly, since I'm traveling:
we decided that the way that teams/"profiles" will work is to have a separate branch (e.g. `licensing-profile`) and push there, so we can use the regular GitHub tools (issues and PRs) to work collaboratively on that.
Once this is stabilized, it will be presented/discussed in a tech call and then merged into  `main` branch.

**Comment by swinslow** [comment]
Thanks @zvr! I'll go ahead and close this PR shortly, then I'll create a new `licensing-profile` branch and will set up a new PR to submit there.

File: model/Licensing/Classes/CopyrightTextField.md
Comment (goneall) [inline model/Licensing/Classes/CopyrightTextField.md:3]: Interesting approach to covering the `NONE` and `NOASSERTION` values.  I don't have an opinion yet on this approach, but it is a bit different than the SPDX 2.X approach.

I'll open a separate issue in the model to discuss a consistent approach for the `NONE` and `NOASSERTIONS`.

File: model/Licensing/Classes/CustomLicense.md
Comment (goneall) [inline model/Licensing/Classes/CustomLicense.md:3]: This is called `ExtractedLicenseInfo` in the 2.X model

File: model/Licensing/Classes/DisjunctiveLicenseSet.md
Comment (goneall) [inline model/Licensing/Classes/DisjunctiveLicenseSet.md:24]: Similar comment to above on difference in naming from 2.X - Note that since `AnyLicenseInfo` is abstract, the impact of changing the name is not that great, but it would be nice to be consistent.

File: model/Licensing/Classes/License.md
Comment (goneall) [inline model/Licensing/Classes/License.md:7]: I have a slightly different conceptual model - a "License Expression" is just text which gets parsed into a logical model which follows more of a tree structure with classes and subclasses.

You might want to take a look at the diagram in the SPDX spec repo and the text in the [SPDX RDF Terms](https://spdx.org/rdf/terms/) for a reference on the 2.X.model.  The class names should line up and you can just copy/paste the descriptions if you agree with the 2.X model.

File: model/Licensing/Classes/LicenseExpression.md
Comment (goneall) [inline model/Licensing/Classes/LicenseExpression.md:3]: Suggest replacing with `AnyLicenseInfo`

File: model/Licensing/Classes/NoAssertionLicense.md
Comment (goneall) [inline model/Licensing/Classes/NoAssertionLicense.md:3]: In RDF in 2.X, this is handled as an Individual rather than a different class.  That being said, what you have here may be a better approach.

Related to the [issue I just submitted](https://github.com/spdx/spdx-3-model/issues/76).

**Review by goneall — COMMENTED** [review]
Hi @swinslow - I finally got a chance to do a more thorough review.

I found some inconsistencies with the [RDF terms](https://spdx.org/rdf/terms/) and [Model from 2.x](https://github.com/spdx/spdx-spec/blob/development/v2.3.1/model/SPDX-UML-Class-Diagram.jpg) which I pointed out.  I didn't catch all of them - There were a few where a term is repeated (e.g. `LicenseExpression` vs `AnyLicenseInfo`) which I only pointed out one or two of the uses in the review comments.

I'd like to go through and do a direct comparison with the RDF terms and propose more consistent text and terms, but I won't have time to do that for a few days.

Since this may create a lot of changes, you could go ahead and just create a new branch, merge these in, and create an issue to sync up with the existing RDF terms.  I can go through and do the work next week and not hold up other changes.

I also noticed a different approach on `NONE` and `NOASSERTION` - which may be better than the 2.X approach.  I opened a separate issue #76 to track the more general question.

**Comment by swinslow** [comment]
@goneall Thanks for this! I'll take a look at your comments in connection with #77, since I'm closing this PR and have opened that one to merge into the new `licensing-profile` branch instead. Appreciate the review.

**Pull request closed without merging**
